src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+                    .compare_and_append(updates.iter(), self.upper.clone(), new_upper.clone())
+                    .await
+                {
+                    Ok(Ok(Ok(()))) => {
+                        trace!("Upper advanced to {new_upper:?}!");
+                        self.upper = new_upper;
+                        break;
+                    }
+                    Ok(Ok(Err(_))) => {
+                        // Upper is already up to date, no need to do any work
+                        break;
+                    }
+                    Ok(Err(invalid_use)) => panic!("compare_and_append failed: {invalid_use}"),
+                    Err(external_err) => {
+                        panic!("compare_and_append failed: {external_err}");
+                    }
+                };
+            }
+        }
+    }
+
+    /// Process a SourceStatusUpdate emitted by a source
+    pub async fn update_status(&mut self, status_update: SourceStatusUpdate) {
+        trace!(
+            "Processing status update: {status_update:?}, current status is {current_status}",
+            current_status = &self.current_status
+        );
+        // Only update status if it is a valid transition
+        if self.active && self.current_status.can_transition(&status_update.status) {
+            loop {
+                let next_ts = (self.now)();
+                let new_upper = Antichain::from_elem(next_ts + 1);
+
+                let updates = self.prepare_row_update(&status_update, next_ts);
+                match self
+                    .write_handle
+                    .compare_and_append(updates.iter(), self.upper.clone(), new_upper.clone())
+                    .await
+                {
+                    Ok(Ok(Ok(()))) => {
+                        trace!("Successful append!");
+                        self.upper = new_upper;
+                        // Update internal status only after a successful append
+                        self.current_status = status_update.status;
+                        break;
+                    }
+                    Ok(Ok(Err(actual_upper))) => {
+                        trace!(
+                            "Had to retry, old upper {:?}, new upper {:?}",
+                            &self.upper,
+                            &actual_upper
+                        );
+                        // Sync to the new upper, go to the loop again
+                        self.sync(&actual_upper.0).await;
+                        // If we can't transition to the new status after the sync, no need to do anything else
+                        if !self.current_status.can_transition(&status_update.status) {
+                            break;
+                        }
+                    }
+                    Ok(Err(invalid_use)) => panic!("compare_and_append failed: {invalid_use}"),
+                    Err(external_err) => {
+                        panic!("compare_and_append failed: {external_err}");
```

Comment:
Same as above, I _feel_ like we could retry on this one if we sync up, like for the `Ok(Ok(Err` case.